### PR TITLE
Support bid submission

### LINKED ISSUE #11: Support bid submission
Construct a bid and submit it to the solver.

### PR DESCRIPTION
**Pull request by shenkeyao**
Closes #11.

### This PR:
* Adds a bid module for bid construction.
* Implements permissioned and non-permissioned versions of event handling for the `ViewFinished` event and bid submission.
* Renames `Types` to `TYPES` for consistency with HotShot.
* Updates dependencies.

### This PR does not:
* Include tests.

### Key places to review:
* `service.rs`.

### CODE REVIEW

File: src/bid.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+pub struct BidConfig {
+    account_seed: [u8; 32],
+    account_index: u64,
+    bid_amount: U256,
+    namespaces: Vec<u32>,
+}
+
+/// Read the bid configuration file.
+pub fn read_bid_config_file(file_path: &str) -> Result<BidConfig, serde_json::Error> {
+    let mut file = File::open(file_path).expect("Failed to open bid config file");
```
Comment (tbro) [inline src/bid.rs:28]: I'm curious, why not just `read_to_string` directly?

```rust
let file = std::fs::read_to_string(file_path).expect("Failed to open bid config file");
from_str(&file)
```

File: src/bid.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+/// Bid configurations can be found via `read_bid_config_file`.
+pub fn from_bid_config(
+    bid_config: BidConfig,
+    view_number: ViewNumber,
+    bid_base_url: Url,
+) -> Result<BidTx, BuildError> {
+    let (account, key) =
+        FeeAccount::generated_from_seed_indexed(bid_config.account_seed, bid_config.account_index);
+    let bid_amount = FeeAmount(bid_config.bid_amount);
+    let url = Url::parse(&format!("{}/{:?}/bundle", bid_base_url, view_number)).map_err(|e| {
```
Comment (elliedavidson) [inline src/bid.rs:46]: The url should be just the base url.  The HotShot leader will add the `bundle` and 'view number` paths to the URL themselves.

**Review by elliedavidson — CHANGES_REQUESTED** [review]
One small comment, otherwise looks good!

File: src/bid.rs (hunk 1 above)
Comment (shenkeyao) [reply src/bid.rs]: This is better. Updated!

File: src/bid.rs (hunk 2 above)
Comment (shenkeyao) [reply src/bid.rs]: Makes sense, fixed.

**Review by elliedavidson — APPROVED** [review]

File: src/service.rs [hunk 3]
```diff
@@ -1121,3 +1198,39 @@ where
...
+    };
+
+    if let Err(e) = solver_client
+        .post::<()>("submit_bid")
+        .body_json(&bid_tx)
+        .unwrap()
+        .send()
+        .await
+    {
+        return Err(BuildError::Error {
```
Comment (QuentinI) [inline src/service.rs:1230]: I don't think we should return an error here, or the builder will exit on any solver failure, that may well be intermittent. I'd say we just log the error here - and maybe add some retry logic, although that can be added in a separate PR.

**Review by QuentinI — CHANGES_REQUESTED** [review]
LGTM in general with a comment on error handling

File: src/service.rs (hunk 3 above)
Comment (shenkeyao) [reply src/service.rs]: Updated the return type. Also updated `handle_received_txns` as discussed on Zulip.

**Review by QuentinI — APPROVED** [review]

**Pull request merged**